File: myintent/src/main/java/pl/mareklangiewicz/myintent/MIContentProvider.java
package pl.mareklangiewicz.myintent;

import android.app.SearchManager;
import android.content.ContentProvider;
import android.content.ContentUris;
import android.content.ContentValues;
import android.content.UriMatcher;
import android.database.Cursor;
import android.database.sqlite.SQLiteDatabase;
import android.net.Uri;
import android.support.annotation.NonNull;

public class MIContentProvider extends ContentProvider {


    private MIDBHelper mMIDBHelper;

    static final UriMatcher sUriMatcher = buildUriMatcher();

    static final int MATCH_CMD_RECENT_DIR = 101;
    static final int MATCH_CMD_RECENT_ITEM = 102;
    static final int MATCH_CMD_EXAMPLE_DIR = 201;
    static final int MATCH_CMD_EXAMPLE_ITEM = 202;
    static final int MATCH_CMD_SUGGEST_DIR_ALL = 305;
    static final int MATCH_CMD_SUGGEST_DIR_LIKE = 306;
    static final int MATCH_RULE_USER_DIR = 401;
    static final int MATCH_RULE_USER_ITEM = 402;

    static UriMatcher buildUriMatcher() {
        final UriMatcher matcher = new UriMatcher(UriMatcher.NO_MATCH);
        matcher.addURI(MIContract.AUTH, MIContract.CmdRecent.PATH, MATCH_CMD_RECENT_DIR);
        matcher.addURI(MIContract.AUTH, MIContract.CmdRecent.PATH + "/#", MATCH_CMD_RECENT_ITEM);
        matcher.addURI(MIContract.AUTH, MIContract.CmdExample.PATH, MATCH_CMD_EXAMPLE_DIR);
        matcher.addURI(MIContract.AUTH, MIContract.CmdExample.PATH + "/#", MATCH_CMD_EXAMPLE_ITEM);
        matcher.addURI(MIContract.AUTH, MIContract.CmdSuggest.PATH, MATCH_CMD_SUGGEST_DIR_ALL);
        matcher.addURI(MIContract.AUTH, MIContract.CmdSuggest.PATH + "/*", MATCH_CMD_SUGGEST_DIR_LIKE);
        matcher.addURI(MIContract.AUTH, MIContract.RuleUser.PATH, MATCH_RULE_USER_DIR);
        matcher.addURI(MIContract.AUTH, MIContract.RuleUser.PATH + "/#", MATCH_RULE_USER_ITEM);
        return matcher;
    }

    public MIContentProvider() {
    }

    @Override
    public boolean onCreate() {
        mMIDBHelper = new MIDBHelper(getContext());
        return true;
    }

    // No need to call this method. This to assist the testing framework in running smoothly.
    // http://developer.android.com/reference/android/content/ContentProvider.html#shutdown()
    @Override
    public void shutdown() {
        mMIDBHelper.close();
        super.shutdown();
    }


    @Override
    public synchronized String getType(@NonNull Uri uri) {
        final int match = sUriMatcher.match(uri);
        switch(match) {
            case MATCH_CMD_RECENT_DIR:
                return MIContract.CmdRecent.TYPE_DIR;
            case MATCH_CMD_RECENT_ITEM:
                return MIContract.CmdRecent.TYPE_ITEM;
            case MATCH_CMD_EXAMPLE_DIR:
                return MIContract.CmdExample.TYPE_DIR;
            case MATCH_CMD_EXAMPLE_ITEM:
                return MIContract.CmdExample.TYPE_ITEM;
            case MATCH_CMD_SUGGEST_DIR_ALL:
            case MATCH_CMD_SUGGEST_DIR_LIKE:
                return MIContract.CmdSuggest.TYPE_DIR;
            case MATCH_RULE_USER_DIR:
                return MIContract.RuleUser.TYPE_DIR;
            case MATCH_RULE_USER_ITEM:
                return MIContract.RuleUser.TYPE_ITEM;
            default:
                throw new UnsupportedOperationException("Unknown uri: " + uri);
        }
    }

    @Override
    public synchronized Cursor query(@NonNull Uri uri, String[] projection, String selection, String[] selectionArgs, String sortOrder) {

        SQLiteDatabase db = mMIDBHelper.getReadableDatabase();
        final int match = sUriMatcher.match(uri);

        String limit = uri.getQueryParameter("limit");

        switch(match) {
            case MATCH_CMD_RECENT_ITEM: // WARNING: selection and selectionArgs are ignored in this case!
                selection = " " + MIContract.CmdRecent._ID + " = ? ";
                selectionArgs = new String[] {uri.getLastPathSegment()};
            case MATCH_CMD_RECENT_DIR:
                return db.query(MIContract.CmdRecent.TABLE_NAME, projection, selection, selectionArgs, null, null, sortOrder, limit);
            case MATCH_CMD_EXAMPLE_ITEM:// WARNING: selection and selectionArgs are ignored in this case!
                selection = " " + MIContract.CmdExample._ID + " = ? ";
                selectionArgs = new String[] {uri.getLastPathSegment()};
            case MATCH_CMD_EXAMPLE_DIR:
                return db.query(MIContract.CmdExample.TABLE_NAME, projection, selection, selectionArgs, null, null, sortOrder, limit);
            case MATCH_CMD_SUGGEST_DIR_LIKE:// WARNING: selection and selectionArgs and sortOrder are ignored in this case!
                selection = SearchManager.SUGGEST_COLUMN_TEXT_1 + " LIKE ? ";
                selectionArgs = new String[] {"%" + Uri.decode(uri.getLastPathSegment()) + "%"};
            case MATCH_CMD_SUGGEST_DIR_ALL:// WARNING: sortOrder is ignored in this case!
                sortOrder = " " + MIContract.CmdSuggest.COL_PRIORITY + " DESC , " + MIContract.CmdSuggest.COL_TIME + " DESC ";
                return db.query(MIContract.CmdSuggest.TABLE_NAME, projection, selection, selectionArgs, null, null, sortOrder, limit);
            case MATCH_RULE_USER_ITEM:// WARNING: selection and selectionArgs are ignored in this case!
                selection = " " + MIContract.RuleUser._ID + " = ? ";
                selectionArgs = new String[] {uri.getLastPathSegment()};
            case MATCH_RULE_USER_DIR:
                return db.query(MIContract.RuleUser.TABLE_NAME, projection, selection, selectionArgs, null, null, sortOrder, limit);
            default:
                throw new UnsupportedOperationException("Unknown uri: " + uri);
        }
    }

    @Override
    public synchronized Uri insert(@NonNull Uri uri, ContentValues values) {

        SQLiteDatabase db = mMIDBHelper.getWritableDatabase();
        final int match = sUriMatcher.match(uri);
        String table;
        switch(match) {
            case MATCH_CMD_RECENT_DIR:
                table = MIContract.CmdRecent.TABLE_NAME;
                break;
            case MATCH_CMD_EXAMPLE_DIR:
                table = MIContract.CmdExample.TABLE_NAME;
                break;
            case MATCH_RULE_USER_DIR:
                table = MIContract.RuleUser.TABLE_NAME;
                break;
            default:
                throw new UnsupportedOperationException("Unknown uri: " + uri);
        }
        long id = db.insert(table, null, values);
        if(id > 0)
            return ContentUris.withAppendedId(uri, id);
        else
            throw new android.database.SQLException("Failed to insert row into " + uri);
    }

    @Override
    public synchronized int delete(@NonNull Uri uri, String selection, String[] selectionArgs) {

        SQLiteDatabase db = mMIDBHelper.getWritableDatabase();
        final int match = sUriMatcher.match(uri);

        switch(match) {
            case MATCH_CMD_RECENT_ITEM:
                selection = " " + MIContract.CmdRecent._ID + " = ? ";
                selectionArgs = new String[] {uri.getLastPathSegment()};
            case MATCH_CMD_RECENT_DIR:
                return db.delete(MIContract.CmdRecent.TABLE_NAME, selection, selectionArgs);
            case MATCH_CMD_EXAMPLE_ITEM:
                selection = " " + MIContract.CmdExample._ID + " = ? ";
                selectionArgs = new String[] {uri.getLastPathSegment()};
            case MATCH_CMD_EXAMPLE_DIR:
                return db.delete(MIContract.CmdExample.TABLE_NAME, selection, selectionArgs);
            case MATCH_RULE_USER_ITEM:
                selection = " " + MIContract.RuleUser._ID + " = ? ";
                selectionArgs = new String[] {uri.getLastPathSegment()};
            case MATCH_RULE_USER_DIR:
                return db.delete(MIContract.RuleUser.TABLE_NAME, selection, selectionArgs);
            default:
                throw new UnsupportedOperationException("Unknown uri: " + uri);
        }
    }

    @Override
    public synchronized int update(@NonNull Uri uri, ContentValues values, String selection, String[] selectionArgs) {

        SQLiteDatabase db = mMIDBHelper.getWritableDatabase();
        final int match = sUriMatcher.match(uri);

        switch(match) {
            case MATCH_CMD_RECENT_ITEM:
                selection = " " + MIContract.CmdRecent._ID + " = ? ";
                selectionArgs = new String[] {uri.getLastPathSegment()};
            case MATCH_CMD_RECENT_DIR:
                return db.update(MIContract.CmdRecent.TABLE_NAME, values, selection, selectionArgs);
            case MATCH_CMD_EXAMPLE_ITEM:
                selection = " " + MIContract.CmdExample._ID + " = ? ";
                selectionArgs = new String[] {uri.getLastPathSegment()};
            case MATCH_CMD_EXAMPLE_DIR:
                return db.update(MIContract.CmdExample.TABLE_NAME, values, selection, selectionArgs);
            case MATCH_RULE_USER_ITEM:
                selection = " " + MIContract.RuleUser._ID + " = ? ";
                selectionArgs = new String[] {uri.getLastPathSegment()};
            case MATCH_RULE_USER_DIR:
                return db.update(MIContract.RuleUser.TABLE_NAME, values, selection, selectionArgs);
            default:
                throw new UnsupportedOperationException("Unknown uri: " + uri);
        }
    }
}

File: myintent/src/main/java/pl/mareklangiewicz/myintent/MIContentProvider.kt
package pl.mareklangiewicz.myintent

import android.app.SearchManager
import android.content.ContentProvider
import android.content.ContentUris
import android.content.ContentValues
import android.content.UriMatcher
import android.database.Cursor
import android.net.Uri
import android.provider.BaseColumns
import pl.mareklangiewicz.myintent.MIContract.AUTH
import pl.mareklangiewicz.myintent.MIContract.CmdExample
import pl.mareklangiewicz.myintent.MIContract.CmdRecent
import pl.mareklangiewicz.myintent.MIContract.CmdSuggest
import pl.mareklangiewicz.myintent.MIContract.RuleUser

class MIContentProvider : ContentProvider() {



    private val MATCH_CMD_RECENT_DIR = 101
    private val MATCH_CMD_RECENT_ITEM = 102
    private val MATCH_CMD_EXAMPLE_DIR = 201
    private val MATCH_CMD_EXAMPLE_ITEM = 202
    private val MATCH_CMD_SUGGEST_DIR_ALL = 305
    private val MATCH_CMD_SUGGEST_DIR_LIKE = 306
    private val MATCH_RULE_USER_DIR = 401
    private val MATCH_RULE_USER_ITEM = 402

    private val matcher = UriMatcher(UriMatcher.NO_MATCH).apply {
        addURI(AUTH, CmdRecent.PATH, MATCH_CMD_RECENT_DIR)
        addURI(AUTH, CmdRecent.PATH + "/#", MATCH_CMD_RECENT_ITEM)
        addURI(AUTH, CmdExample.PATH, MATCH_CMD_EXAMPLE_DIR)
        addURI(AUTH, CmdExample.PATH + "/#", MATCH_CMD_EXAMPLE_ITEM)
        addURI(AUTH, CmdSuggest.PATH, MATCH_CMD_SUGGEST_DIR_ALL)
        addURI(AUTH, CmdSuggest.PATH + "/*", MATCH_CMD_SUGGEST_DIR_LIKE)
        addURI(AUTH, RuleUser.PATH, MATCH_RULE_USER_DIR)
        addURI(AUTH, RuleUser.PATH + "/#", MATCH_RULE_USER_ITEM)
    }

    private lateinit var dbhelper: MIDBHelper

    override fun onCreate(): Boolean {
        dbhelper = MIDBHelper(context!!)
        return true
    }

    // No need to call this method. This to assist the testing framework in running smoothly.
    // http://developer.android.com/reference/android/content/ContentProvider.html#shutdown()
    override fun shutdown() {
        dbhelper.close()
        super.shutdown()
    }


    @Synchronized override fun getType(uri: Uri): String? {
        return when (matcher.match(uri)) {
            MATCH_CMD_RECENT_DIR -> CmdRecent.TYPE_DIR
            MATCH_CMD_RECENT_ITEM -> CmdRecent.TYPE_ITEM
            MATCH_CMD_EXAMPLE_DIR -> CmdExample.TYPE_DIR
            MATCH_CMD_EXAMPLE_ITEM -> CmdExample.TYPE_ITEM
            MATCH_CMD_SUGGEST_DIR_ALL, MATCH_CMD_SUGGEST_DIR_LIKE -> CmdSuggest.TYPE_DIR
            MATCH_RULE_USER_DIR -> RuleUser.TYPE_DIR
            MATCH_RULE_USER_ITEM -> RuleUser.TYPE_ITEM
            else -> throw UnsupportedOperationException("Unknown uri: " + uri)
        }
    }


    @Synchronized override fun query(uri: Uri, projection: Array<String>?, selection: String?, selectionArgs: Array<String>?, sortOrder: String?): Cursor? {

        val db = dbhelper.readableDatabase
        val match = matcher.match(uri)

        val limit = uri.getQueryParameter("limit")

        return when (match) {

             // WARNING: selection and selectionArgs are ignored in this case!
            MATCH_CMD_RECENT_ITEM -> db.query(
                    CmdRecent.TABLE_NAME, projection,
                    " ${BaseColumns._ID} = ? ", arrayOf(uri.lastPathSegment),
                    null, null, sortOrder, limit
            )

            MATCH_CMD_RECENT_DIR -> return db.query(
                    CmdRecent.TABLE_NAME, projection,
                    selection, selectionArgs,
                    null, null, sortOrder, limit
            )

             // WARNING: selection and selectionArgs are ignored in this case!
            MATCH_CMD_EXAMPLE_ITEM -> db.query(
                    CmdExample.TABLE_NAME, projection,
                    " ${BaseColumns._ID} = ? ", arrayOf(uri.lastPathSegment),
                    null, null, sortOrder, limit
            )

            MATCH_CMD_EXAMPLE_DIR -> db.query(
                    CmdExample.TABLE_NAME, projection,
                    selection, selectionArgs,
                    null, null, sortOrder, limit
            )

            // WARNING: selection and selectionArgs and sortOrder are ignored in this case!
            MATCH_CMD_SUGGEST_DIR_LIKE -> db.query(
                    CmdSuggest.TABLE_NAME, projection,
                    "${SearchManager.SUGGEST_COLUMN_TEXT_1} LIKE ? ", arrayOf("%${Uri.decode(uri.lastPathSegment)}%"),
                    null, null, " ${CmdSuggest.COL_PRIORITY} DESC , ${CmdSuggest.COL_TIME} DESC ", limit
            )

             // WARNING: sortOrder is ignored in this case!
            MATCH_CMD_SUGGEST_DIR_ALL -> db.query(
                    CmdSuggest.TABLE_NAME, projection,
                    selection, selectionArgs,
                    null, null, " ${CmdSuggest.COL_PRIORITY} DESC , ${CmdSuggest.COL_TIME} DESC ", limit
            )

             // WARNING: selection and selectionArgs are ignored in this case!
            MATCH_RULE_USER_ITEM -> db.query(
                    RuleUser.TABLE_NAME, projection,
                    " ${BaseColumns._ID} = ? ", arrayOf(uri.lastPathSegment),
                    null, null, sortOrder, limit
            )

            MATCH_RULE_USER_DIR -> db.query(
                    RuleUser.TABLE_NAME, projection,
                    selection, selectionArgs,
                    null, null, sortOrder, limit
            )

            else -> throw UnsupportedOperationException("Unknown uri: $uri")
        }
    }

    @Synchronized override fun insert(uri: Uri, values: ContentValues?): Uri? {

        val db = dbhelper.writableDatabase
        val table = when (matcher.match(uri)) {
            MATCH_CMD_RECENT_DIR -> CmdRecent.TABLE_NAME
            MATCH_CMD_EXAMPLE_DIR -> CmdExample.TABLE_NAME
            MATCH_RULE_USER_DIR -> RuleUser.TABLE_NAME
            else -> throw UnsupportedOperationException("Unknown uri: $uri")
        }
        val id = db.insert(table, null, values)
        if (id > 0)
            return ContentUris.withAppendedId(uri, id)
        else
            throw android.database.SQLException("Failed to insert row into $uri")
    }



    @Synchronized override fun delete(uri: Uri, selection: String?, selectionArgs: Array<String>?): Int {

        val db = dbhelper.writableDatabase

        return when (matcher.match(uri)) {
            MATCH_CMD_RECENT_ITEM -> db.delete(CmdRecent.TABLE_NAME, " ${BaseColumns._ID} = ? ", arrayOf(uri.lastPathSegment))
            MATCH_CMD_RECENT_DIR -> db.delete(CmdRecent.TABLE_NAME, selection, selectionArgs)
            MATCH_CMD_EXAMPLE_ITEM -> db.delete(CmdExample.TABLE_NAME, " ${BaseColumns._ID} = ? ", arrayOf(uri.lastPathSegment))
            MATCH_CMD_EXAMPLE_DIR -> db.delete(CmdExample.TABLE_NAME, selection, selectionArgs)
            MATCH_RULE_USER_ITEM -> db.delete(RuleUser.TABLE_NAME, " ${BaseColumns._ID} = ? ", arrayOf(uri.lastPathSegment))
            MATCH_RULE_USER_DIR -> return db.delete(RuleUser.TABLE_NAME, selection, selectionArgs)
            else -> throw UnsupportedOperationException("Unknown uri: $uri")
        }
    }

    @Synchronized override fun update(uri: Uri, values: ContentValues?, selection: String?, selectionArgs: Array<String>?): Int {

        val db = dbhelper.writableDatabase

        return when (matcher.match(uri)) {
            MATCH_CMD_RECENT_ITEM -> db.update(CmdRecent.TABLE_NAME, values, " ${BaseColumns._ID} = ? ", arrayOf(uri.lastPathSegment))
            MATCH_CMD_RECENT_DIR -> db.update(CmdRecent.TABLE_NAME, values, selection, selectionArgs)
            MATCH_CMD_EXAMPLE_ITEM -> db.update(CmdExample.TABLE_NAME, values, " ${BaseColumns._ID} = ? ", arrayOf(uri.lastPathSegment))
            MATCH_CMD_EXAMPLE_DIR -> db.update(CmdExample.TABLE_NAME, values, selection, selectionArgs)
            MATCH_RULE_USER_ITEM -> db.update(RuleUser.TABLE_NAME, values, " ${BaseColumns._ID} = ? ", arrayOf(uri.lastPathSegment))
            MATCH_RULE_USER_DIR -> db.update(RuleUser.TABLE_NAME, values, selection, selectionArgs)
            else -> throw UnsupportedOperationException("Unknown uri: $uri")
        }
    }
}

